### PR TITLE
Refactored metrics ingestion from druid

### PR DESCRIPTION
**Pull request by Chester-Ong**
Changelog
1. Prometheus **pod** parameter => **host**
2. Reverse DNS lookup ToPodName() returns initial value if lookup fails
    + from localhost:8888 => "" to localhost:8888 => "localhost:8888
3. Returns the Port too under host

### CODE REVIEW

**Comment by exherb** [comment]
why remove podNames? dns reverse lookup is not stable.

**Comment by Chester-Ong** [comment]
correct me if i'm wrong but isn't podnames kubernetes specific infrastructure. Furthermore, if I map hostname to podNames it would be quite puzzling for a normal user who isn't using kubernetes.

For the reverse dns, I would presume that it interacts with k8 dns service. The only logic I've changed from the old code to revert to the input host if reverse dns lookup fails. Do you expect an empty string if the lookup fails for your use case?

**Comment by iamabhishek-dubey** [comment]
@exherb would you like to test this PR? Otherwise, I can validate things in k8s as well. Also, I think @Chester-Ong point is valid that this is not specific to k8s only :)

**Comment by exherb** [comment]
I add this logic as ip-domain cache because when pod is in terminal status, the reverse dns query will fail.

**Comment by exherb** [comment]
maybe podNames should rename to ipToHosts. If you think it’s confuse

**Comment by iamabhishek-dubey** [comment]
I guess this should be fine but we have to test it on both k8s and non-k8s environment.

**Comment by Chester-Ong** [comment]
@exherb I added a global dnsCache using go-cache as the backend would this caching strategy work for you?

I don't think ipToHost would be a good idea. It is a good idea if you need to inform the metrics consumer how did you parse **druid.hostname** but I would suspect that most consumers do not care how did you obtain the **host** value. Hence, a much more identifiable host would be easier to understand.

@iamabhishek-dubey it is your call on the final structure, whether you would like host, ipToHost, podName or any combination on the 3.

Lastly, I would like to mention that this is my first go project so code reviews will be much appreciated.

File: collector/interface.go [hunk 1]
```diff
@@ -73,11 +75,17 @@ func (w worker) podName() string {
 	return strings.Split(w.Worker.IP, ".")[0]
 }
```
Comment (iamabhishek-dubey) [inline collector/interface.go:76]: I guess this method and struct is not getting used anywhere, Can you see if we can use this to get worker information as well?

File: collector/interface.go [hunk 2]
```diff
@@ -73,11 +75,17 @@ func (w worker) podName() string {
...
 	names, err := net.LookupAddr(host)
 	if err != nil || len(names) == 0 {
-		return ""
+		dnsCache.Set(host, host, cache.DefaultExpiration)
+		return host
 	}
-	return strings.Split(names[0], ".")[0]
+	dnsResult := strings.Split(names[0], ".")[0]
+	dnsCache.Set(host, dnsResult, cache.DefaultExpiration)
+	return dnsResult
```
Comment (iamabhishek-dubey) [inline collector/interface.go:90]: Let's move this function to `utils`

File: listener/druid_endpoint.go [hunk 3]
```diff
@@ -32,53 +30,46 @@ func DruidHTTPEndpoint(gauge *prometheus.GaugeVec) http.HandlerFunc {
...
 			for _, data := range druidData {
-				if data["dataSource"] != nil {
-					metricName := fmt.Sprintf("%v", data["metric"])
-					serviceName := fmt.Sprintf("%v", data["service"])
-					host := fmt.Sprintf("%v", data["host"])
-					value, _ := strconv.ParseFloat(fmt.Sprintf("%v", data["value"]), 64)
+				metric := data["metric"].(string)
+				service := data["service"].(string)
+				hostname := data["host"].(string)
+				value, _ := data["value"].(float64)
```
Comment (iamabhishek-dubey) [inline listener/druid_endpoint.go:36]: Let's handle the error here using logger

File: collector/interface.go (hunk 1 above)
Comment (exherb) [reply collector/interface.go:76]: rename to hostname

Comment (Chester-Ong) [reply collector/interface.go:76]: its being used in 
```
func (collector *MetricCollector) Collect(ch chan<- prometheus.Metric)
...
    podName = worker.podName()
```

Comment (Chester-Ong) [reply collector/interface.go:74]: I don't get the worker information part of your question.
I renamed the function to hostname to be consistent with the new naming convention.

File: collector/interface.go (hunk 2 above)
Comment (Chester-Ong) [reply collector/interface.go]: Moved

File: listener/druid_endpoint.go (hunk 3 above)
Comment (Chester-Ong) [reply listener/druid_endpoint.go:36]: added logging, run with **--log.level trace**

**Comment by exherb** [comment]
LGTM, thanks

**Pull request merged**